### ISSUE TITLE
*Possible* bug in P2P (device-to-device) copy overlapped with work on GPU

### ISSUE THREAD

**Issue by yuribd** [pytorch/pytorch#15568]
## 🐛 Bug
 
Following very simple code prints value of 1 to output (tensor(1., device=‘cuda:1’)); instead of 0, i’d expect to see. Using default stream (instead of custom) ‘fixes’ the issue, but defeats the purpose of original code - copy from one GPU to several others at the same time (async), also overlapping with (unrelated) computations on these GPUs. *Would be happy for any other solutions which achieve this goal?* (likely i'm just doing it in a wrong way?)

Can only guess that something odd happens with memory when “x1 + 1” starts executing (see nvprof) and at the same time device-to-device copy starts happening

cross-referenced from https://discuss.pytorch.org/t/possible-bug-in-p2p-device-to-device-copy-overlapped-with-work-on-gpu/33033

## To Reproduce
```python
import torch

n = int(3e4)

d0 = torch.device(‘cuda:0’)
x0 = torch.zeros(n, n, device=d0)

d1 = torch.device(‘cuda:1’)
x1 = torch.zeros(n, n, device=d1)
x1 + 1
x1 + 1

s = torch.cuda.Stream()
with torch.cuda.stream(s):
    x2 = x0.to(d1)

torch.cuda.synchronize() # just in case, x2[0,0] should sync anyway; sync on device-1 (d1) doesn’t help, using events don’t help either, neither does time.sleep() - looks like x2 is already 'corrupted' by this point (see nvprof) 

print(x2[0, 0])
```

## Expected behavior

x2 should not be 'corrupted' - i.e. all elements should be zeros

## Environment
 - PyTorch Version 1.0
 - OS Linux
 - built from src (as conda build seem not to have OMP multithreading? - separate issue)
 - Python 3.7.1 
 - CUDA/cuDNN version: 10
 - GPU models V100

- Same issue happens using old pytorch version (0.4.1) from conda distribution, and CUDA 9 and different type of GPU - which likely means it's a 'feature' rather than bug (yet surprising). Sorry if so, please feel free to close.

---
nvprof output
```
Start Duration Grid Size Block Size Regs* SSMem* DSMem* Size Throughput SrcMemType DstMemType Device Context Stream Src Dev Src Ctx Dst Dev Dst Ctx Name
6.69668s 2.0160us - - - - - 3.3528GB 2e+06GB/s Device - Tesla V100-SXM2 1 7 - - - - [CUDA memset]
13.0215s 1.9520us - - - - - 3.3528GB 2e+06GB/s Device - Tesla V100-SXM2 2 18 - - - - [CUDA memset]
13.0261s 4.7312ms (878907 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - ZN2at6native18elementwise_kernelILi512ELi1EZNS0_16gpu_unary_kernelIZNS0_17gpu_binary_kernelIZNS0_15add_kernel_implIfEEvRNS_14TensorIteratorEN3c106ScalarEEUlffE_EEvS6_RKT_EUlfE0_EEvS6_SC_EUliE_EEviT1 [886]
13.0292s 74.185ms - - - - - 3.3528GB 45.194GB/s Device Device Tesla V100-SXM2 1 28 Tesla V100-SXM2 1 Tesla V100-SXM2 2 [CUDA memcpy PtoP]
13.0308s 5.0902ms (878907 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - ZN2at6native18elementwise_kernelILi512ELi1EZNS0_16gpu_unary_kernelIZNS0_17gpu_binary_kernelIZNS0_15add_kernel_implIfEEvRNS_14TensorIteratorEN3c106ScalarEEUlffE_EEvS6_RKT_EUlfE0_EEvS6_SC_EUliE_EEviT1 [889]
13.0359s 5.0761ms (878907 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - ZN2at6native18elementwise_kernelILi512ELi1EZNS0_16gpu_unary_kernelIZNS0_17gpu_binary_kernelIZNS0_15add_kernel_implIfEEvRNS_14TensorIteratorEN3c106ScalarEEUlffE_EEvS6_RKT_EUlfE0_EEvS6_SC_EUliE_EEviT1 [902]
13.0410s 5.0793ms (878907 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - ZN2at6native18elementwise_kernelILi512ELi1EZNS0_16gpu_unary_kernelIZNS0_17gpu_binary_kernelIZNS0_15add_kernel_implIfEEvRNS_14TensorIteratorEN3c106ScalarEEUlffE_EEvS6_RKT_EUlfE0_EEvS6_SC_EUliE_EEviT1 [905]
13.1045s 2.6560us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply3<TensorEQOp<float, unsigned char>, unsigned char, float, float, unsigned int, int=1, int=1, int=1>(OffsetInfo<unsigned char, float, float>, OffsetInfo<TensorEQOp<float, unsigned char>, float, unsigned int>, OffsetInfo<unsigned char, float, int=1>, float, float) [1014]
13.1046s 1.5040us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply2<Tensor_abs_Float_Op, float, float, unsigned int, int=1, int=1>(OffsetInfo<float, unsigned int, int=1>, OffsetInfo<float, unsigned int, int=1>, unsigned int, Tensor_abs_Float_Op) [1028]
13.1047s 1.7280us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply2<TensorNEValueOp<float, unsigned char>, unsigned char, float, unsigned int, int=1, int=1>(OffsetInfo<unsigned char, unsigned char, float>, OffsetInfo<TensorNEValueOp<float, unsigned char>, unsigned char, unsigned int>, unsigned char, float) [1040]
13.1048s 1.6630us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply3<TensorBitAndOp, unsigned char, unsigned char, unsigned char, unsigned int, int=1, int=1, int=1>(OffsetInfo<TensorBitAndOp, unsigned char, unsigned int>, OffsetInfo<unsigned char, unsigned char, int=1>, OffsetInfo<unsigned char, unsigned char, int=1>, unsigned char, unsigned char) [1051]
13.1048s 1.1840us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply2<TensorNEValueOp<float, unsigned char>, unsigned char, float, unsigned int, int=1, int=1>(OffsetInfo<unsigned char, unsigned char, float>, OffsetInfo<TensorNEValueOp<float, unsigned char>, unsigned char, unsigned int>, unsigned char, float) [1062]
13.1048s 1.2160us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply3<TensorBitAndOp, unsigned char, unsigned char, unsigned char, unsigned int, int=1, int=1, int=1>(OffsetInfo<TensorBitAndOp, unsigned char, unsigned int>, OffsetInfo<unsigned char, unsigned char, int=1>, OffsetInfo<unsigned char, unsigned char, int=1>, unsigned char, unsigned char) [1073]
13.1049s 3.6480us (1 1 1) (1024 1 1) 30 0B 8.0000KB - - - - Tesla V100-SXM2 2 18 - - - - void kernelReduceAll<unsigned char, unsigned int, long, thrust::identity, ReduceAdd, int=1>(TensorInfo<unsigned char, unsigned int>, unsigned int, long, long, thrust::identity, long*) [1086]
13.1049s 3.2000us - - - - - 8B 2.3842MB/s Device Pageable Tesla V100-SXM2 2 18 - - - - [CUDA memcpy DtoH]
13.1051s 1.9520us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - ZN2at4cuda75_GLOBAL__N__51_tmpxft_000055a3_00000000_11_Copy_compute_75_cpp1_ii_dd3fb9a321kernelPointwiseApply2IZN75_GLOBAL__N__51_tmpxft_000055a3_00000000_11_Copy_compute_75_cpp1_ii_dd3fb9a36CopyOpIlhE5applyERNS_6TensorERKS6_EUlRlRKhE_lhjLi1ELi1ELi1EEEvNS0_6detail10TensorInfoIT0_T2_EENSF_IT1_SH_EESH_T [1101]
13.1051s 1.3440us (1 1 1) (128 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void thrust::cuda_cub::core::_kernel_agent<thrust::cuda_cub::__scan::InitAgent<thrust::cuda_cub::cub::ScanTileState<long, bool=1>, int>, thrust::cuda_cub::cub::ScanTileState<long, bool=1>, int>(bool=1, thrust::cuda_cub::cub::ScanTileState<long, bool=1>) [1121]
13.1051s 7.0080us (1 1 1) (96 1 1) 42 0B 2.0781KB - - - - Tesla V100-SXM2 2 18 - - - - void thrust::cuda_cub::core::_kernel_agent<thrust::cuda_cub::__scan::ScanAgent<thrust::device_ptr, thrust::device_ptr, thrust::plus, int, long, thrust::detail::integral_constant<bool, bool=0>>, thrust::device_ptr, thrust::device_ptr, thrust::plus, int, thrust::cuda_cub::cub::ScanTileState<long, bool=1>, thrust::cuda_cub::__scan::AddInitToExclusiveScan<long, thrust::plus>>(thrust::device_ptr, thrust::device_ptr, long, thrust::plus, int, long) [1126]
13.1052s 1.7600us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply3<TensorMaskedSelectOp<float, unsigned char, long>, unsigned char, long, float, unsigned int, int=1, int=1, int=1>(OffsetInfo<unsigned char, unsigned char, long>, OffsetInfo<long, unsigned char, float>, OffsetInfo<TensorMaskedSelectOp<float, unsigned char, long>, unsigned char, unsigned int>, unsigned char, float) [1132]
13.1052s 1.2480us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply2<Tensor_abs_Float_Op, float, float, unsigned int, int=1, int=1>(OffsetInfo<float, unsigned int, int=1>, OffsetInfo<float, unsigned int, int=1>, unsigned int, Tensor_abs_Float_Op) [1146]
13.1053s 1.8240us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - ZN2at4cuda75_GLOBAL__N__51_tmpxft_000055a3_00000000_11_Copy_compute_75_cpp1_ii_dd3fb9a321kernelPointwiseApply2IZN75_GLOBAL__N__51_tmpxft_000055a3_00000000_11_Copy_compute_75_cpp1_ii_dd3fb9a36CopyOpIdfE5applyERNS_6TensorERKS6_EUlRdRKfE_dfjLi1ELi1ELi1EEEvNS0_6detail10TensorInfoIT0_T2_EENSF_IT1_SH_EESH_T [1159]
13.1054s 5.6320us (1 1 1) (1024 1 1) 16 0B 8.0000KB - - - - Tesla V100-SXM2 2 18 - - - - void kernelReduceAll<double, unsigned int, double, thrust::identity, ReduceMin, int=1>(TensorInfo<double, unsigned int>, unsigned int, double, double, thrust::identity, double*) [1172]
13.1054s 1.8880us - - - - - 8B 4.0410MB/s Device Pageable Tesla V100-SXM2 2 18 - - - - [CUDA memcpy DtoH]
13.1055s 1.3440us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply1<TensorFillOp, double, unsigned int, int=1>(OffsetInfo<TensorFillOp, double, unsigned int>, double, double) [1184]
13.1055s 5.4080us (1 1 1) (1024 1 1) 16 0B 8.0000KB - - - - Tesla V100-SXM2 2 18 - - - - void kernelReduceAll<double, unsigned int, double, thrust::identity, ReduceMax, int=1>(TensorInfo<double, unsigned int>, unsigned int, double, double, thrust::identity, double*) [1196]
13.1055s 1.5040us - - - - - 8B 5.0727MB/s Device Pageable Tesla V100-SXM2 2 18 - - - - [CUDA memcpy DtoH]
13.1056s 1.1840us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply1<TensorFillOp, double, unsigned int, int=1>(OffsetInfo<TensorFillOp, double, unsigned int>, double, double) [1208]
13.1057s 1.5680us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply2<Tensor_ceil_Float_Op, float, float, unsigned int, int=1, int=1>(OffsetInfo<float, unsigned int, int=1>, OffsetInfo<float, unsigned int, int=1>, unsigned int, Tensor_ceil_Float_Op) [1226]
13.1057s 1.5680us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply3<TensorNEOp<float, unsigned char>, unsigned char, float, float, unsigned int, int=1, int=1, int=1>(OffsetInfo<unsigned char, float, float>, OffsetInfo<TensorNEOp<float, unsigned char>, float, unsigned int>, OffsetInfo<unsigned char, float, int=1>, float, float) [1238]
13.1058s 1.6000us - - - - - 1B 610.35KB/s Device Pageable Tesla V100-SXM2 2 18 - - - - [CUDA memcpy DtoH]
13.1059s 2.3680us (1 1 1) (128 1 1) 26 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - ZN2at6native18elementwise_kernelILi128ELi4EZNS0_17gpu_binary_kernelIZNS0_15div_kernel_implIdEEvRNS_14TensorIteratorEEUlddE_EEvS5_RKT_EUliE0_EEviT1 [1259]
13.1060s 1.5360us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply2<TensorGTValueOp<double, unsigned char>, unsigned char, double, unsigned int, int=1, int=1>(OffsetInfo<unsigned char, unsigned char, double>, OffsetInfo<TensorGTValueOp<double, unsigned char>, unsigned char, unsigned int>, unsigned char, double) [1270]
13.1060s 1.4720us - - - - - 1B 663.42KB/s Device Pageable Tesla V100-SXM2 2 18 - - - - [CUDA memcpy DtoH]
13.1060s 1.2480us (1 1 1) (512 1 1) 16 0B 0B - - - - Tesla V100-SXM2 2 18 - - - - void kernelPointwiseApply2<TensorGTValueOp<double, unsigned char>, unsigned char, double, unsigned int, int=1, int=1>(OffsetInfo<unsigned char, unsigned char, double>, OffsetInfo<TensorGTValueOp<double, unsigned char>, unsigned char, unsigned int>, unsigned char, double) [1289]
13.1060s 1.4720us - - - - - 1B 663.42KB/s Device Pageable Tesla V100-SXM2 2 18 - - - - [CUDA memcpy DtoH]
13.1061s 1.4720us - - - - - 4B 2.5915MB/s Device Pageable Tesla V100-SXM2 2 18 - - - - [CUDA memcpy DtoH]
13.1061s 1.9520us - - - - - 4B 1.9543MB/s Device Pageable Tesla V100-SXM2 2 18 - - - - [CUDA memcpy DtoH]
```

**Comment by yuribd** [comment]
Many thanks for having a look!! Yes, that’s what i initially tried, but then it doesn’t do async copy to 2 devices: d0 to d1 and d0 to d2 (this part is not in somified code i posted; i’ll provide example with nvprof soon)-it only does them async if all done in stream belonging to d0 (source arr, rather than one stream per destination device)

**Comment by yuribd** [comment]
If all streams belong to d0 then all (computation on GPU-0, transfer from 0 to 1 and transfer from 0 to 2) is done in parallel (async) – start around 21.580s (ses nvprof)

```python
import torch

n = int(3e4)    

d0 = torch.device('cuda:0')
x0 = torch.zeros(n, n, device=d0)

d1 = torch.device('cuda:1')
x1 = torch.zeros(n, n, device=d1)

d2 = torch.device('cuda:2')
x2 = torch.zeros(n, n, device=d2)

s0 = torch.cuda.Stream()
s01 = torch.cuda.Stream()
s02 = torch.cuda.Stream()

torch.cuda.synchronize()
with torch.cuda.device(d1):
	torch.cuda.synchronize()
with torch.cuda.device(d2):
	torch.cuda.synchronize()


with torch.cuda.stream(s0):
	x0 @ x0

with torch.cuda.stream(s01):
    x1 = x0.to(d1)

with torch.cuda.stream(s02):
    x2 = x0.to(d2)       

torch.cuda.synchronize()
with torch.cuda.device(d1):
	torch.cuda.synchronize()
with torch.cuda.device(d2):
	torch.cuda.synchronize()
```
```
nvprof
   Start  Duration            Grid Size      Block Size     Regs*    SSMem*    DSMem*      Size  Throughput  SrcMemType  DstMemType           Device   Context    Stream          Src Dev   Src Ctx          Dst Dev   Dst Ctx  Name
7.44163s  1.6640us                    -               -         -         -         -  3.3528GB   2e+06GB/s      Device           -  Tesla V100-SXM2         1         7                -         -                -         -  [CUDA memset]
14.5811s  1.9520us                    -               -         -         -         -  3.3528GB   2e+06GB/s      Device           -  Tesla V100-SXM2         2        18                -         -                -         -  [CUDA memset]
21.5654s  1.6320us                    -               -         -         -         -  3.3528GB   2e+06GB/s      Device           -  Tesla V100-SXM2         3        29                -         -                -         -  [CUDA memset]
21.5738s  1.9840us                    -               -         -         -         -      112B  53.836MB/s    Pageable      Device  Tesla V100-SXM2         1         7                -         -                -         -  [CUDA memcpy HtoD]
21.5751s  3.53039s          (235 235 1)       (256 1 1)       118  16.250KB        0B         -           -           -           -  Tesla V100-SXM2         1        39                -         -                -         -  volta_sgemm_128x128_nn [1148]
21.5800s  74.248ms                    -               -         -         -         -  3.3528GB  45.157GB/s      Device      Device  Tesla V100-SXM2         1        41  Tesla V100-SXM2         1  Tesla V100-SXM2         2  [CUDA memcpy PtoP]
21.5855s  74.185ms                    -               -         -         -         -  3.3528GB  45.195GB/s      Device      Device  Tesla V100-SXM2         1        43  Tesla V100-SXM2         1  Tesla V100-SXM2         3  [CUDA memcpy PtoP]
```

However changing streams to destination devices forces it to sync {copy from d0 to d1} and {copy from d0 to d2} -2d starts exactly 74ms after 1st

```python
s01 = torch.cuda.Stream(d1)
s02 = torch.cuda.Stream(d2)
```

adding ‘with torch.cuda.device’ (for d0->d1 and d0->d2 copy) doesn’t change the picture either:
```python
with torch.cuda.device(d1):
	with torch.cuda.stream(s01):
	    x1 = x0.to(d1)
```
```
   Start  Duration            Grid Size      Block Size     Regs*    SSMem*    DSMem*      Size  Throughput  SrcMemType  DstMemType           Device   Context    Stream          Src Dev   Src Ctx          Dst Dev   Dst Ctx  Name
12.3526s  1.6320us                    -               -         -         -         -  3.3528GB   2e+06GB/s      Device           -  Tesla V100-SXM2         1         7                -         -                -         -  [CUDA memset]
19.4457s  1.9520us                    -               -         -         -         -  3.3528GB   2e+06GB/s      Device           -  Tesla V100-SXM2         2        18                -         -                -         -  [CUDA memset]
26.2154s  1.6000us                    -               -         -         -         -  3.3528GB   2e+06GB/s      Device           -  Tesla V100-SXM2         3        29                -         -                -         -  [CUDA memset]
26.2250s  2.0160us                    -               -         -         -         -      112B  52.982MB/s    Pageable      Device  Tesla V100-SXM2         1         7                -         -                -         -  [CUDA memcpy HtoD]
26.2260s  3.52558s          (235 235 1)       (256 1 1)       118  16.250KB        0B         -           -           -           -  Tesla V100-SXM2         1        39                -         -                -         -  volta_sgemm_128x128_nn [1282]
26.2305s  74.248ms                    -               -         -         -         -  3.3528GB  45.156GB/s      Device      Device  Tesla V100-SXM2         1         7  Tesla V100-SXM2         1  Tesla V100-SXM2         2  [CUDA memcpy PtoP]
26.3048s  74.185ms                    -               -         -         -         -  3.3528GB  45.195GB/s      Device      Device  Tesla V100-SXM2         1         7  Tesla V100-SXM2         1  Tesla V100-SXM2         3  [CUDA memcpy PtoP]
```

**Comment by ngimel** [comment]
The first (failing) example is a result of caching allocator not properly tracking free blocks because the stream is on another device. Printouts and very slight modifications added to the first example:
```
import torch

n = int(3e4)

d0 = torch.device('cuda:0')
x0 = torch.zeros(n, n, device=d0)

d1 = torch.device('cuda:1')
x1 = torch.zeros(n, n, device=d1)
print("0", x1.data_ptr())
x1 + 2
a = x1 + 3
print("a", a.data_ptr())
del a

s = torch.cuda.Stream()
with torch.cuda.stream(s):
    x2 = x0.to(d1)


torch.cuda.synchronize() # just in case, x2[0,0] should sync anyway; sync on device-1 (d1) doesn’t help, using events don't help either, neither does time.sleep() - looks like x2 is already 'corrupted' by this point (see nvprof) 
print("x2", x2.data_ptr(), x2.device)

print(x2[0, 0], x2[-1,-1])
```
outputs
```
0 140302824243200
a 140299200364544
x2 140299200364544 cuda:1
tensor(3., device='cuda:1') tensor(0., device='cuda:1')
```
What happens is  `x2` is using the same memory as `a` because stream dependencies are not properly captured, and `a=x1+3` overwrites already copied part of x2. data_ptr of `a` is the same as data_ptr of x2, the value in x2 is actually the result of the second operation (`a=x1+3`), last value in x2 is ok because it is written after `a=x1+3` is completed. 
cc @colesbury for caching allocator

**Comment by ngimel** [comment]
> However changing streams to destination devices forces it to sync {copy from d0 to d1} and {copy from d0 to d2} -2d starts exactly 74ms after 1st

Yes, that's expected based on the comment @soumith linked to, copies are always put on a current stream on a source device, so unless you are explicitly setting different streams on the source device for each copy, copies will be synchronized.

**Comment by colesbury** [comment]
I might be missing something, but this looks like an issue in the copy kernel. The stream `s` is on device 0, not device 1. The copy should logically "happen" in `s` on device 0 and the default stream in device 1. As in, other operations in the default stream on device 1 should be sequenced correctly w.r.t. the copy.

**Comment by ngimel** [comment]
There is no correct or incorrect sequencing of pw operations on device 1, because they don't depend on copy in any way, and can overlap with copy, but the issue is that third kernel on this profile (which corresponds to `a=x1+3`, each pw operation is split into 2 kernels by TensorIterator) gets the same output pointer on d1 as copy operation, and overwrites its result. I'm not sure if its a bug, but it's certainly unexpected result.
![selection_049](https://user-images.githubusercontent.com/15841449/50614972-86725080-0e97-11e9-8d98-fb89efb6159c.png)

**Comment by colesbury** [comment]
This isn't an issue with the caching allocator. For example, you can move the x2 construction outside the stream and have the same issue:

https://gist.github.com/colesbury/7c40453b3ffe249bd9945954cf742ffc

It's a limitation of the copy code. Specifically, if the copy occurs on a non-default stream it does no extra synchronization on the destination device:

https://github.com/pytorch/pytorch/blob/bb3c3f516bceca922e9c334960fc7ca20aba54f1/aten/src/ATen/native/cuda/Copy.cu#L70-L73

So you need to manually synchronize with the default stream on device 1 before the copy. You also nee d a synchronization after the copy to be correct.

I think we should change the copy code so that it behaves as if it occurred in both the src and dst stream *always* instead of only if the src stream is the default stream.

There are also a few related bugs:
1) In the example code, it's reading `x0` from `s` without synchronizing from s to device 0's default stream
2) The PyTorch Python (not C++) stream API is often incorrect for the default stream on the non-current device.

**Comment by ezyang** [comment]
> I think we should change the copy code so that it behaves as if it occurred in both the src and dst stream always instead of only if the src stream is the default stream.

This sounds good to me.

**Comment by yuribd** [comment]
Many thanks for having a look!
To sum up, is there any way *currently* to achieve what second example does, but without side effect of 'memory corruption' - "If all streams belong to d0 then all (computation on GPU-0, transfer from 0 to 1 and transfer from 0 to 2) is done in parallel (async) – start around 21.580s (ses nvprof)"

**Comment by yuribd** [comment]
"In the example code, it's reading x0 from s without synchronizing from s to device 0's default stream"

Yes, sorry; in fact originally had sync there (and do have it for 2d example), but decided to remove to simplify code as 'corruption' doesn't originate there (from what i see in nvprof, memset finished before any p2p copy if i read it correctly) and adding it doesn't fix the issue

**Comment by yuribd** [comment]
@ngimel 

> Yes, that's expected based on the comment @soumith linked to, copies are always put on a current stream on a source device, so unless you are explicitly setting different streams on the source device for each copy, copies will be synchronized.

Thanks for confirming! So seems i'm doing it sort of right way, except didn't yet get how *not* to 'corrupt memory' on destination devices  and at the same time avoid costly sync of all work on destination device prior to copy

**Comment by colesbury** [comment]
> To sum up, is there any way currently to achieve what second example does, but without side effect of 'memory corruption'?:

Yes, although it's complicated further because of https://github.com/pytorch/pytorch/issues/15682.

Here's the code with necessary inter-stream synchronization:

https://gist.github.com/colesbury/e83a0b0f1f0b3e46c84490fc9b420e34

It's tricky and took me a while to get right. I'm honestly not 100% confident in the code's correctness, which is a sign that this is way too hard. Fixing #15682 and the changing the copy behavior should help.

**Comment by colesbury** [comment]
@ngimel points out that the above example doesn't overlap copy with compute on device 1. Here's a modification with that overlaps copy with compute:

https://gist.github.com/colesbury/d6cb564b522e4d7ace623eef763770bf

**Comment by yuribd** [comment]
Many thanks, @colesbury !!

**Comment by yuribd** [comment]
I see, thanks! Guess alternative solution (haven't tried yet) would be to schedule work on d1 in *non*-default stream?

**Comment by yuribd** [comment]
> That's provided we don't do any work in default stream on d1

And it seems to be surprisingly non-trivial :-( Looks like i'm facing this issue currently - https://github.com/pytorch/pytorch/pull/8354

Sorry for so many questions, but is there a way to force autograd to run on non-default stream? It would make code (in my case) so much easier

**Comment by ezyang** [comment]
Unfortunately I don't think we have any facility for doing so. If you don't mind editing code it might not be too hard to hack up `torch/csrc/autograd/engine.cpp` to run on a different stream, but then you're on the hook for making sure synchronization happens correctly.

**Comment by ngimel** [comment]
You can have an AutogradFunction for changing streams, along the lines of https://gist.github.com/ngimel/55267aeeb74586e67ea466fc1ed6fc11
Note that the last thing of your forward pass should be changing stream, otherwise backward won't be putting backward work on the stream you want. Also, the correct stream has to be set at the beginning of each forward iteration

**Comment by ngimel** [comment]
>In addition, it gives speed up when there are more than one model (which are not related to each other) are trained on one GPU, as they can run in completely separate streams (backward pass not clashing on default stream)
Careful with that, because cudnn and cublas handles are not thread safe, so if you are trying to run your models from the different threads, there will be synchronization issues. cudnn handle pool that fixes this issue has been recently merged, but cublas is still unsafe.

**Comment by yuribd** [comment]
Thank you so much, @ngimel , @ezyang, @colesbury, @soumith !

**Comment by yuribd** [comment]
Hopefully last quesion: in @colesbury solution, why do we need to wait for `s1` ? https://gist.github.com/colesbury/d6cb564b522e4d7ace623eef763770bf#file-copy-py-L27

**Comment by yuribd** [comment]
sorry, what i meant is it *permanently* changes current stream on device-1 *not-respecting* context manager

**Comment by yuribd** [comment]
to sum-up, solution suggested by @colesbury (with slightly modification for my case) seem to work fine as far as i can tell so far (nvvp, etc)

**Comment by mrshenli** [comment]
> If the copy is not on the default stream, then it is up to the user to add needed synchronization on the dst device, since the stream on the dst device that wishes to synchronize may not be the same index as the one on the src device. 

Hi @colesbury, I am trying to understand the comments you highlighted above. Why src and dst stream indices mismatch prevents us from synchronizing on the dst stream?

**Comment by mrshenli** [comment]
@soumith @yuribd @ngimel @colesbury @ezyang 
Has anyone started editing `Copy.cu` to solve the src/dst stream wait problem? If not, I am happy to work it.
